services/kubernetesmisc/src/handlers/kubernetesBuildCancel.ts

```diff
@@ -0,0 +1,87 @@
+import { getOpenShiftInfoForProject, updateTask } from '@lagoon/commons/src/api';
+import { logger } from "@lagoon/commons/src/local-logging";
+import { sendToLagoonLogs } from '@lagoon/commons/src/logs';
+
+import Api, { ClientConfiguration } from 'kubernetes-client';
+const Client = Api.Client1_13;
+
+const ocsafety = string => string.toLocaleLowerCase().replace(/[^0-9a-z-]/g, '-');
+
+const generateSanitizedNames = (project, environment, projectInfo) => {
+  try {
+    const safeBranchName = ocsafety(environment.name);
+    const safeProjectName = ocsafety(project.name);
+    const namespace = projectInfo.openshiftProjectPattern
+      ? projectInfo.openshiftProjectPattern
+          .replace('${branch}', safeBranchName)
+          .replace('${project}', safeProjectName)
+      : `${safeProjectName}-${safeBranchName}`;
+    return { namespace, safeProjectName };
+  } catch (error) {
+    logger.error(`Error while loading information for project ${project.name}`);
+    logger.error(error);
+    throw error;
+  }
+};
+
+const getUrlTokenFromProjectInfo = (projectOpenShift, name) => {
+  try {
+    const url = projectOpenShift.openshift.consoleUrl.replace(/\/$/, '');
+    const token = projectOpenShift.openshift.token || '';
+    return { url, token };
+  } catch (error) {
+    logger.warn(
+      `Error while loading information for project ${name}: ${error}`
+    );
+    throw error;
+  }
+};
+
+const getConfig = (url, token) => ({
+  url,
+  insecureSkipTlsVerify: true,
+  auth: {
+    bearer: token
+  }
+});
+
+const deleteJob = async (client: Api.ApiRoot, namespace: string, jobName: string) => {
+  try {
+    const options = {
+      body: {
+        kind: 'DeleteOptions',
+        apiVersion: 'v1',
+        propagationPolicy: 'Foreground',
+      }
+    };
+    // https://github.com/godaddy/kubernetes-client/blob/master/docs/1.13/Job.md#apisbatchv1namespacesnamespacejobsnamedelete
+    return await client.apis.batch.v1.namespaces(namespace).jobs(jobName).delete(options)
+  } catch (err) {
+    logger.error(`Couldn't delete job ${jobName}. Error: ${err}`);
+  }
+};
+
+const kubernetesBuildCancel = async (data: any) => {
+  const { build: { name: buildName }, project, environment } = data;
+
+  const { project: projectInfo } = await getOpenShiftInfoForProject(project.name);
+  const { url, token } = getUrlTokenFromProjectInfo(projectInfo, project.name);
+  const config: ClientConfiguration = getConfig(url, token);
+  const client = new Client({ config });
+  const { namespace, safeProjectName } = generateSanitizedNames(project, environment, projectInfo);
+
+  const deleteResult = deleteJob(client, namespace, buildName);
```

Comment:
OpenShift builds are a full "product" and can handle edge cases like cancelling a build that's already failed or completed. Since we're using jobs, we probably should double check that the job is still active right before we cancel it.

I also wonder how we should convey the status to the API. If we just delete the job, the `kubernetesjobsmonitor` will see the job is missing and automatically mark it as "failed" with no log messages. That's probably OK for now? The API does have a `CANCELLED` status type though which we could update here. We could also add a log message "cancelled by user/lagoon api." We'd have to somehow tell the jobs monitor that the missing job is expected and not to mark it failed.